service/src/util/serviceUtils.ts

```diff
@@ -87,3 +87,37 @@ export async function modifyResourcesForDraft(artifacts: FhirArtifact[], version
 
   return artifacts;
 }
+
+/**
+ * Helper function that takes an active or draft artifact and returns it with a new id,
+ * same status, url (from the $clone parameter), and version (from the $clone parameter).
+ */
+export async function modifyResourcesForClone(artifacts: FhirArtifact[], version: string) {
+  for (const artifact of artifacts) {
+    artifact.id = uuidv4();
+    artifact.version = version;
+    if (artifact.relatedArtifact) {
+      artifact.relatedArtifact.forEach(ra => {
+        if (
+          ra.type === 'composed-of' &&
+          ra.resource &&
+          ra.extension?.some(
+            e => e.url === 'http://hl7.org/fhir/StructureDefinition/artifact-isOwned' && e.valueBoolean === true
+          )
+        ) {
+          const oldUrl = ra.resource.split('|')[0];
+          ra.resource = oldUrl + '-clone|' + version;
+        }
+      });
+    }
+
+    const checkExisting = await findArtifactByUrlAndVersion(artifact.url, version, artifact.resourceType);
```

Comment:
Doesn't the database already check uniqueness by url and version when inserting a new artifact?

Reply:
I think it only does by id? Happy to discuss more

Reply:
https://github.com/projecttacoma/measure-repository/blob/f876096accea79defc79c03a8a0c8842421c51e2/service/scripts/dbSetup.ts#L21 has a url/version index as well.

Reply:
Did a quick look at removing the checkExisting bits, and the server appropriately sends the duplicate key error: 
"Unexpected: E11000 duplicate key error collection: measure-repository.Measure index: url_1_version_1 dup key: { url: \"http://example1.com\", version: \"1.0.0\" }"
If we want to do a Bad Request error instead, it might be cleaner to catch the database error and rethrow a Bad Request from it.

Reply:
@hossenlopp - maybe you have thoughts on this?

Reply:
Yeah. This error should be caught and turned into a bad request. If we can identify the error we could give feedback that this resource already exists.

Reply:
The `findArtifactByUrlAndVersion` function throws a BadRequest error to give feedback that this resource already exists and this happens before the batch insertion happens. Do you want me to move this to `batchClone` and `batchDraft` (where those database insertions are occurring)? Those insertions are already in a try catch so should I check that it is a duplicate key error specifically?

Reply:
I'm thinking that a duplicate key error would typically be a bad request across the board (in all cases, not just this one, but interested if other folks agree), so maybe we can strip out the checks here, leave it as a straight duplicate key for now, and create a new task for turning duplicate key into bad request error across the board.

Reply:
Yeah I think that's a good idea considering we currently don't wrap any duplicate key errors in a BadRequestError.